### PR TITLE
Search styling overlay

### PR DESCRIPTION
**Pull request by budzanowski**
Create a styling overlay over Search component input field.

This is a div positioned over input that allows to insert other elements like for example span with CSS with background modifications that will style the text.

The Search component has now a new prop `overlyaStyling`. You can pass here a function that will receive current Search input content and based on that string it will generate styling.

Below is possible example where sting in format `taxonomy:term` is highlighted. 

image

Code for this highlight is also committed in themes-magic-search-card.jsx - this is intermediate work for new style Search in Themes Showcase. It is not visible in production - hidden after a flag and it is committed to accelerate development. Effects are visible in `/design` for dev and wpcalypso.

### CODE REVIEW

File: client/components/search/index.jsx [hunk 1]
```diff
@@ -161,6 +164,18 @@ const Search = React.createClass( {
 		}
 	},
 
+	scrollOverlay: function() {
+		const _this = this;
+
+		window.requestAnimationFrame(function() {
+			console.log( 'init_tokens: ' + _this.refs.overlay.scrollLeft );
```
Comment (seear) [inline client/components/search/index.jsx]: Need to remove the debug code here (or change it to logging)

File: client/components/search/index.jsx [hunk 2]
```diff
@@ -161,6 +164,18 @@ const Search = React.createClass( {
 		}
 	},
 
+	scrollOverlay: function() {
+		const _this = this;
```
Comment (seear) [inline client/components/search/index.jsx]: What is this line for?

File: client/components/search/index.jsx [hunk 3]
```diff
@@ -312,7 +329,7 @@ const Search = React.createClass( {
 				</div>
 				<div className={ fadeDivClass }>
 					<input
-						type="search"
+						type="text"
```
Comment (seear) [inline client/components/search/index.jsx]: How will this affect existing instances of the `<Search/>` component?

File: client/my-sites/themes/themes-magic-search-card/index.jsx [hunk 4]
```diff
@@ -65,6 +69,24 @@ const ThemesMagicSearchCard = React.createClass( {
 		this.setState( { searchIsOpen: false } );
 	},
 
+	searchTokens( input ) {
+		const tokens = input.split(/(\s+)/);
+		let cls;
```
Comment (seear) [inline client/my-sites/themes/themes-magic-search-card/index.jsx]: `cls` confused me initially. `className` might be better. Though maybe I'm just being slow this morning :)

File: client/components/search/index.jsx [hunk 5]
```diff
@@ -52,6 +52,7 @@ const Search = React.createClass( {
 		onSearchOpen: PropTypes.func,
 		onSearchClose: PropTypes.func,
 		analyticsGroup: PropTypes.string,
+		overlayStyling: PropTypes.func,
```
Comment (seear) [inline client/components/search/index.jsx]: We should probably explain this new prop in the docs.

**Comment by seear** [comment]
@budzanowski In 5e419bb I've made the overlay text transparent, which means that the scroll does not have to be so accurate because we never see the shadow text. This lets us keep `type=search`, avoiding disruption to existing `<Search/>` instances.

Does that make sense, or was there another reason for `type=text`?

**Comment by budzanowski** [comment]
It is hard to tell. I have observed that `type=text` has no additional styling in browsers where `type=search` has and in it is browser dependent. For my browser and taste the difference where not acceptable especially when cursor was in different place where it should be. There are more behavior and styling differences but they are not affecting use. `type=search` can change styling with next iterations of browser engines. I have used the overlay text to detect mismatch in positioning. In the end it needs to be removed. The mismatched text will also make the overlay styling not aligned with actual text which will be more visible where more advanced styling for token will be implemented. I think that for now we can leave it as `type=search` and decide to change when it will turn out that it is as source of problems.

**Comment by seear** [comment]
> I think that for now we can leave it as type=search and decide to change when it will turn out that it is as source of problems.

Yeah, sounds good.

File: client/components/search/index.jsx (hunk 5 above)
Comment (seear) [inline client/components/search/index.jsx]: Added

**Comment by seear** [comment]
Ok, this should be ready to merge now 👍

**Comment by budzanowski** [comment]
Looks good. :+1:

**Pull request merged**